Merge users_items into groups_attempts

zenovich opened the pull request:
Fixes #286

smadbe commented on db/migrations/60_merge_users_items_into_groups_attempts.sql:
```diff
@@ -145,7 +145,8 @@ ALTER TABLE `users_items`
     ADD COLUMN `platform_data_removed` tinyint(4) NOT NULL DEFAULT '0'
-        AFTER `answer`;
+        AFTER `answer`,
+    ADD INDEX `ancestors_computation_state` (`ancestors_computation_state`);
```
Ah, my test of migrate-undo is not totally useless ;-)

smadbe commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+										 JOIN items ON(
+										items.ID = items_items.child_item_id
+								)
+						WHERE items.type <> 'Course' AND items.no_score = 0
```
This ` items.type <> 'Course'` can be removed. It will be a default of the UI.

zenovich replied:
I don't get what you mean. A child never can be a course?

zenovich replied:
If so, it is fixed.

smadbe replied:
Sorry, my explanation was a bit short. A child can be a course, but no need to consider that a course means "no_score", they will set "no_score" to true on course if they need so. (instead of implicitly)

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+					WITH task_children_data_view AS (
+						WITH best_scores AS (
+							SELECT group_id, item_id, MAX(score) AS score, MAX(validated) AS validated,
+								MAX(validated_at) AS validated_at
```
https://github.com/France-ioi/AlgoreaPlatform/blob/master/shared/listeners.php#L56 They have MAX there. Should I change it to MIN anyway?

zenovich replied:
If so, should I also change MAX to MIN when choosing `validated_at` among different item children?

smadbe replied:
That's not the same, that should probably be max among items, but min across attempts.

zenovich replied:
fixed

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
 				const updateQuery = `
-					UPDATE users_items
+					WITH task_children_data_view AS (
+						WITH best_scores AS (
```
fixed

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+										 JOIN items_items ON(
+										parent_groups_attempts.item_id = items_items.parent_item_id
+								)
+										 LEFT JOIN best_scores AS task_children ON(
```
fixed

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+						)
+						SELECT
+								parent_groups_attempts.id,
+								SUM(IF(task_children.group_id IS NOT NULL AND task_children.validated, 1, 0)) AS children_validated,
```
fixed

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+						SELECT
+								parent_groups_attempts.id,
+								SUM(IF(task_children.group_id IS NOT NULL AND task_children.validated, 1, 0)) AS children_validated,
+								SUM(IF(task_children.group_id IS NOT NULL AND task_children.validated, 0, 1)) AS children_non_validated,
```
Not the same actually, but I simplified it too. This subquery JOINs items_items so it is NULL when there are no children, but it is 1 when there are no attempts for a child which is correct.

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+										 JOIN items ON(
+										items.ID = items_items.child_item_id
+								)
+						WHERE items.type <> 'Course' AND items.no_score = 0
```
fixed

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+								SUM(IF(task_children.group_id IS NOT NULL AND task_children.validated, 1, 0)) AS children_validated,
+								SUM(IF(task_children.group_id IS NOT NULL AND task_children.validated, 0, 1)) AS children_non_validated,
+								SUM(IF(items_items.category = 'Validation' AND
+											 (ISNULL(task_children.group_id) OR task_children.validated != 1), 1, 0)) AS children_category,
```
No, it isn't equivalent to that, but I simplified it a bit.

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
 							SUM(children.tasks_solved) AS tasks_solved,
 							SUM(validated) AS children_validated,
-							children.user_id AS user_id,
+							children.group_id AS group_id,
```
fixed

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -98,60 +98,89 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
-							children_data.user_id IS NOT NULL AND items_items.id IS NOT NULL,
-							children_data.children_validated, users_items.children_validated),
-						users_items.validated = IF(task_children_data.user_item_id IS NOT NULL AND items_items.id IS NOT NULL,
+						groups_attempts.latest_activity_at = IF(task_children_data.id IS NOT NULL AND
```
It is a hack for setting groups_attempts.ancestors_computation_state = 'done' in the same query (and keeping all other columns unchanged when there where no children found for the item). In PHP they had a separate query for setting groups_attempts.ancestors_computation_state.

smadbe replied:
Ok, I get why it is equivalent to what was done on the users_items in PHP. But I still don't get the point of not changing the former value when no children. For instance, if its last child has been removed, I don't see why it should store the former "validated" from his child for instance. I think it is more something forgotten before than a "feature". Do you see any reasons?

And in any case, I do not see the benefit of rejoining items_items

zenovich replied:
Leaves never have children and I doubt we want to wipe leaves' data with zeros.

zenovich replied:
For `And in any case, I do not see the benefit of rejoining items_items`, it's fixed now.

smadbe replied:
Here the propagation goes upwards, so it is not about reaching leaves but chapter etc with only "no_score" children. 
I asked the question twice, you haven't found reasons from what I understood. 
So, if we do not find reason (I don't say there isn't), I think we may put more appropriate values when there are no children. In addition to fixing a potential problem, it may make the request more understandable. So I would say:
```sql 
target_groups_attempts.latest_activity_at =
  GREATEST(
       IFNULL(children_stats.latest_activity_at, target_groups_attempts.latest_activity_at),
       IFNULL(target_groups_attempts.latest_activity_at, children_stats.latest_activity_at)
   ), -- GREATEST which returns the other value if one is null
target_groups_attempts.tasks_tried = IFNULL(children_stats.tasks_tried, 0),
target_groups_attempts.tasks_with_help = IFNULL(children_stats.tasks_with_help, 0),
target_groups_attempts.tasks_solved = IFNULL(children_stats.tasks_solved, 0),
target_groups_attempts.children_validated = IFNULL(children_stats.children_validated, 0),
target_groups_attempts.validated = CASE ... END,
target_groups_attempts.validated_at = 
  IF(items.validation_type = 'Categories', children_stats.max_validated_at_categories, children_stats.max_validated_at),
target_groups_attempts.ancestors_computation_state = 'done'
```
... which is what makes more sense to me for propagation across attempts. (slightly different from what was in php)

zenovich replied:
> Here the propagation goes upwards, so it is not about reaching leaves but chapter etc with only "no_score" children.

zenovich replied:
The proposed query would wipe `task_tried`, `tasks_with_help`, `tasks_solved`, `validated`, `validated_at` of leaves with zeros/NULLs.

zenovich commented:
> Now every users_items have been moved to groups_attempts, for the active_attempt_id NULL, the active_attempt_id should be put to this new created record. As it should be only one row, it is just a basic UPDATE to do. Then you should put "active_attempt_id" column to not NULL as well... (I see no row failing on that). (And create the FK? :-))

The foreign key will break the itemTaskTokenGet service :(

zenovich commented:
> Now every users_items have been moved to groups_attempts, for the active_attempt_id NULL, the active_attempt_id should be put to this new created record. As it should be only one row, it is just a basic UPDATE to do. 

Don't get what's the reason of doing this, but I added the UPDATE into the migration anyway. :/

zenovich commented:
Note for `ComputeAllUserItems: transform the propagation of scores through users_items by propagation through groups_attempts (may require creating them if not existing). PropagateAttempts should disappear then.`: ComputeAllUserItems doesn't propagate scores and never did.

smadbe commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -17,10 +17,10 @@ type groupItemPair struct {
-// For users_items marked with ancestors_computation_state = 'todo':
-// 1. We mark all their ancestors in users_items as 'todo'
-//  (we consider a row in users_items as an ancestor if it has the same value in user_id and
+// ComputeAllUserItems recomputes fields of groups_attempts
```
And this one?

zenovich replied:
done

smadbe commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -93,65 +93,79 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+									children_stats.max_validated_at_categories, children_stats.max_validated_at)
+							), groups_attempts.validated_at),
+						groups_attempts.ancestors_computation_state = 'done'
+					WHERE groups_attempts.ancestors_computation_state = 'processing'`
```
And related to your initial "INSERT .. ON DUPLICATE UPDATE ...", so you propagate considering the parent attempt always exists? Is it a wise assumption?

smadbe replied:
?

zenovich replied:
Yes, I presuppose that all the parent attempts needed for propagation are pre-created. The migration creates absent groups_attempt rows according to existing users_items.

smadbe commented:
> > Now every users_items have been moved to groups_attempts, for the active_attempt_id NULL, the active_attempt_id should be put to this new created record. As it should be only one row, it is just a basic UPDATE to do.
> 
> Don't get what's the reason of doing this, but I've added the UPDATE into the migration anyway. :/

The goal of the users_items.active_attempt_id is too have a default attempt to load for the user. Of course, it has to be nullable while this items didn't have any attempt for the user, but now there is one attempt for every users_items, it should be referenced. 
Actually, for the moment, `users_items` only exists to hold `active_attempt_id` so it is strange not to consider as normal to have it not null (except if the existence of the record itself would mean something, which is not the case here from what I can think of).

> Either making the column not nullable or adding the foreign key would break itemTaskTokenGet & itemActiveAttemptUpdate :(

Ok but maybe they would break because they are not adapted to the change as they should. :-) As it will hit later if it is the case, let me check...

For `itemTaskTokenGet`:
```
//   * If there is no row for the current user and the given item in `users_items`, the service creates one.
//
//   * If the active attempt (`active_attempt_id`) is not set in the `users_items` for the item and the user,
//   the service chooses the most recent one among all the user's attempts (or the team's attempts if
//   `items.has_attempts`=1)  for the given item. If no attempts found, the new one gets created and chosen as active.
```
Actually, there is no reason a users_items would have been created if there was no attempt yet. So "no row for the current user and the given item in `users_items`" is equivalent to the former "no row for the current user and the given item in `users_items`" OR " If the active attempt (`active_attempt_id`) is not set in the `users_items`".
So the code, should change, but not that much: if there are no users_items matching, just create an attempt (as now) and then create the missing `users_items`.

For `updateActiveAttempt`, you just have to update or create if not existing the users_items. 

Actually, for both service, there was this `userItemStore.CreateIfMissing(...)` required anyway. Here is can do it is smarter way probably (without the need for the get).

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -93,65 +93,79 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+							GROUP BY group_id, item_id
+						) AS aggregated_children_attempts ON 1
+						JOIN items ON(
+							items.ID = items_items.child_item_id
```
yes, it is :)

zenovich replied:
fixed

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -93,65 +93,79 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
+						JOIN items_items ON(
+							parent_groups_attempts.item_id = items_items.parent_item_id
+						)
+						LEFT JOIN LATERAL (
```
I started from this, but MySQL didn't use indexes in that case :(

smadbe replied:
Worth a comment in the query to mention that.

zenovich replied:
added the comment

zenovich replied:
If we do `LEFT JOIN (SELECT ... GROUP BY ...) ON some_condition` (as suggested above), MySQL would group all the data (for all possible (group_id, item_id) pairs) for each of the outer rows. This means it would perform `(SELECT ... GROUP BY ...)` for each of groups_attempts rows.

zenovich replied:
For the 'wrong results', I just mean that there is now way to get correct results effectively using 'LEFT JOIN' here.

zenovich commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -93,65 +93,79 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
-					UPDATE users_items
-					LEFT JOIN (
+					UPDATE groups_attempts
+					LEFT JOIN LATERAL (
```
Fixed

smadbe commented on app/database/user_item_store_compute_all_user_items.go:
```diff
@@ -108,11 +108,12 @@ func (s *UserItemStore) ComputeAllUserItems() (err error) {
-						FROM items_items
+							MAX(IF(items_items_with_scores.category = 'Validation', aggregated_children_attempts.validated_at, NULL))
+								AS max_validated_at_categories
+						FROM items_items AS items_items_with_scores
```
I don't get this alias. At the scope you use it, it still have the score as you filter it inside this part of the query. 
It is like you would call `items` `items_with_score` while you filter it afterwards.

zenovich replied:
fixed

smadbe commented on db/migrations/60_merge_users_items_into_groups_attempts.sql:
```diff
@@ -31,6 +31,11 @@ FROM `users_items`
+    FROM `users_items`
+    LEFT JOIN `items` ON `items`.`id` = `users_items`.`item_id`
+    WHERE `items`.`id` IS NULL;
+
```
So your "JOIN `items` ON `items`.`id` = `users_items`.`item_id`" in the next update which just there to exclude non existing items (right?) can disappear. 
I think doing the same with `users` may make sense as well (even if I see no rows matching in the current dump)

zenovich replied:
fixed

smadbe commented on app/api/currentuser/get_dump.feature:
```diff
@@ -57,19 +57,21 @@ Feature: Export the short version of the current user's data
       | 113 | 1        | 404     | 0     |
+    And the database has the following table 'users_items':
+      | user_id | item_id | active_attempt_id |
+      | 2       | 404     | 111               |
```
I see in all these changes of features file that it is partially to define `users_items` after `groups_attempts`. 
I think that will quickly be boring to think about FK when writing such dataset, so probably there should be a `SET FOREIGN_KEY_CHECKS = 0` before setting all these tables (set to 1 when ? when doing the request 😕).

zenovich replied:
By not doing this way, I can be sure that all the needed data are there

smadbe replied:
Ah when `SET FOREIGN_KEY_CHECKS = 1`, it would not recheck everything? Make sense. There should be able to recheck everything. 
Anyway, when we'll cross FK (e.g. item have a default string, string are related to item), we'll have to do this.

zenovich commented:
> > > Now every users_items have been moved to groups_attempts, for the active_attempt_id NULL, the active_attempt_id should be put to this new created record. As it should be only one row, it is just a basic UPDATE to do.
> > 
> > 
> > Don't get what's the reason of doing this, but I've added the UPDATE into the migration anyway. :/
> 
> The goal of the users_items.active_attempt_id is too have a default attempt to load for the user. Of course, it has to be nullable while this items didn't have any attempt for the user, but now there is one attempt for every users_items, it should be referenced.
> Actually, for the moment, `users_items` only exists to hold `active_attempt_id` so it is strange not to consider as normal to have it not null (except if the existence of the record itself would mean something, which is not the case here from what I can think of).
> 
> > Either making the column not nullable or adding the foreign key would break itemTaskTokenGet & itemActiveAttemptUpdate :(
> 
> Ok but maybe they would break because they are not adapted to the change as they should. :-) As it will hit later if it is the case, let me check...
> 
> For `itemTaskTokenGet`:
> 
> ```
> //   * If there is no row for the current user and the given item in `users_items`, the service creates one.
> //
> //   * If the active attempt (`active_attempt_id`) is not set in the `users_items` for the item and the user,
> //   the service chooses the most recent one among all the user's attempts (or the team's attempts if
> //   `items.has_attempts`=1)  for the given item. If no attempts found, the new one gets created and chosen as active.
> ```
> 
> Actually, there is no reason a users_items would have been created if there was no attempt yet. So "no row for the current user and the given item in `users_items`" is equivalent to the former "no row for the current user and the given item in `users_items`" OR " If the active attempt (`active_attempt_id`) is not set in the `users_items`".
> So the code, should change, but not that much: if there are no users_items matching, just create an attempt (as now) and then create the missing `users_items`.
> 
> For `updateActiveAttempt`, you just have to update or create if not existing the users_items.
> 
> Actually, for both service, there was this `userItemStore.CreateIfMissing(...)` required anyway. Here is can do it is smarter way probably (without the need for the get).

fixed

zenovich commented on app/api/items/save_grade.feature:
```diff
@@ -51,7 +46,12 @@ Feature: Save grading result
+    And the database has the following table 'users_items':
+      | user_id | item_id | active_attempt_id |
+      | 10      | 50      | 100               |
+      | 10      | 60      | 101               |
```
I don't get what you mean here. The first row is the attempt that should be update and the second one is needed to check that the service doesn't update other attempts.

smadbe replied:
A few lines above we have 
```
And the database has the following table 'groups_attempts':
      | id  | group_id | item_id | hints_requested        | order |	      | id  | group_id | item_id | hints_requested        | order |
      | 100 | 101      | 50      | [0,  1, "hint" , null] | 0     |	      | 100 | 101      | 50      | [0,  1, "hint" , null] | 1     |
      | 101 | 101      | 50      | [0,  1, "hint" , null] | 2     |

```
Meaning the attempt 101 is on (group,item) (101,50)... you cannot have then a users_items with active attempt 101 and item 60. 
(related to the discussion on composite FK)

zenovich replied:
fixed

zenovich commented on db/migrations/61_make_users_items_active_attempt_id_not_null.sql:
```diff
@@ -0,0 +1,30 @@
+-- +migrate Up
+ALTER TABLE `users_items`
```
It is done in `db/migrations/60_merge_users_items_into_groups_attempts.sql`

zenovich commented on db/migrations/61_make_users_items_active_attempt_id_not_null.sql:
```diff
@@ -0,0 +1,30 @@
+    ADD PRIMARY KEY `fk_user_id_item_id` (`user_id`, `item_id`),
+    ADD FOREIGN KEY `fk_user_id` (`user_id`) REFERENCES `users`(`id`) ON DELETE CASCADE,
+    ADD FOREIGN KEY `fk_item_id` (`item_id`) REFERENCES `items`(`id`) ON DELETE CASCADE,
+    ADD FOREIGN KEY `fk_active_attempt_id` (`active_attempt_id`) REFERENCES `groups_attempts` (`id`) ON DELETE CASCADE;
```
I don't see the logic behind `(active_attempt_id, item_id)` since `active_attempt_id` is unique

smadbe replied:
In a nutshell: Otherwise you write a users_items with an `item_id` not matching the one of the `active_attempt_id` and vice-versa. A valid alternative of course would be to remove `item_id` as it is implicit in the active attempt.

zenovich replied:
Can we think about this later (when we'll have FKs in groups_attempts)?

smadbe replied:
Yes we can 🙂

zenovich commented on db/migrations/60_merge_users_items_into_groups_attempts.sql:
```diff
@@ -23,15 +27,43 @@ SELECT
+    ) AS groups_to_insert ON `groups_to_insert`.`id` IS NOT NULL
+    LEFT JOIN `groups_attempts` AS `existing_attempts`
+        ON `existing_attempts`.`group_id` = `groups_to_insert`.`id` AND `existing_attempts`.`item_id` = `items`.`id`
+    WHERE `existing_attempts`.`id` IS NULL;
```
They create a users_items row for the (idUser_id, idItemTeam) pair on contest opening/item starting/team joining (see https://github.com/France-ioi/AlgoreaPlatform/blob/b9441048e604367136b4c45b101c14cf4b85a404/teams/teamsCommon.php#L169), so we can be sure that this row exists in all the cases we need. At the same time, it looks like we cannot be sure that someone didn't modify groups.idTeamItem by hand.

smadbe replied:
But it is also that this code is the latest version of the php code (some things that I thought via php code are not valid at all in the actual data), it has be checked by hand on the actual data.

zenovich commented on db/migrations/60_merge_users_items_into_groups_attempts.sql:
```diff
@@ -23,15 +27,43 @@ SELECT
     `users_items`.`best_answer_at`,`users_items`.`latest_answer_at`,`users_items`.`latest_hint_at`,
-    `users_items`.`ranked`,`users_items`.`all_lang_prog`
+    `users_items`.`ranked`,`users_items`.`all_lang_prog`,
+    IF(`users`.`self_group_id` = `groups_to_insert`.`id`, 'done', 'todo')
```
We recompute stats for every item having at least one descendant marked as 'todo'. So should work fine (I believe).

smadbe replied:
I don't think it works as you would have: 
- one leaf of single-user participation  -> copied -> marked as done: ok
- the parent contest done alone -> copied -> done: ok
- the parent chapter -> copied -> done: this node may be made of 2 contests, one the user has done alone (above) and one the user has done in team. In groups_attempts related to the self_group_id, you should only have results of the user alone, if you recompute, it will be correct but what you have copied is not.
Right? 🤔

zenovich replied:
right. Fixed
